sdks/python/apache_beam/transforms/environments.py

```diff
@@ -0,0 +1,258 @@
+from apache_beam.utils import proto_utils
+
+__all__ = ['Environment',
+           'DockerEnvironment', 'ProcessEnvironment', 'ExternalEnvironment',
+           'PortableEnvironment', 'RunnerAPIEnvironmentHolder']
+
+PORTABLE_ENVIRONMENT = 'portable'
+
+
+class Environment(object):
+
+  _known_urns = {}
+  _urn_to_env_cls = {}
+
+  urn = None
+
+  def to_runner_api_parameter(self, context):
+    raise NotImplementedError
+
+  @classmethod
+  def register_urn(cls, urn, parameter_type, constructor=None):
+
+    def register(constructor):
+      cls._known_urns[urn] = parameter_type, constructor
+      return staticmethod(constructor)
+
+    if constructor:
+      # Used as a statement.
+      register(constructor)
+    else:
+      # Used as a decorator.
+      return register
+
+  @classmethod
+  def register_environment(cls, env_cls):
+    cls._urn_to_env_cls[env_cls.urn] = env_cls
+    return env_cls
+
+  @classmethod
+  def get_env_cls_from_urn(cls, urn):
+    return cls._urn_to_env_cls[urn]
+
+  def to_runner_api(self, context):
+    from apache_beam.portability.api import beam_runner_api_pb2
+    urn, typed_param = self.to_runner_api_parameter(context)
+    return beam_runner_api_pb2.Environment(
+      urn=urn,
+      payload=typed_param.SerializeToString()
+        if isinstance(typed_param, message.Message)
+        else typed_param.encode('utf-8') if isinstance(typed_param, str)
+        else typed_param
+    )
+
+  @classmethod
+  def from_runner_api(cls, proto, context):
+    if proto is None or not proto.urn:
+      return None
+    parameter_type, constructor = cls._known_urns[proto.urn]
+
+    try:
+      return constructor(
+          proto_utils.parse_Bytes(proto.payload, parameter_type),
+          context)
+    except Exception:
+      if context.allow_proto_holders:
+        return RunnerAPIEnvironmentHolder(proto)
+      raise
+
+  @classmethod
+  def from_options(cls, options):
+    raise NotImplementedError
+
+
+@Environment.register_environment
+class DockerEnvironment(Environment):
+
+  urn = common_urns.environments.DOCKER.urn
+
+  def __init__(self, container_image=None):
+    from apache_beam.runners.portability.portable_runner import PortableRunner
+
+    if container_image:
+      self.container_image = container_image
+    else:
+      self.container_image = PortableRunner.default_docker_image()
+
+  def __eq__(self, other):
+    return self.__class__ == other.__class__ \
+           and self.container_image == other.container_image
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.DOCKER.urn,
+            beam_runner_api_pb2.DockerPayload(
+                container_image=self.container_image)
+            )
+  @Environment.register_urn(common_urns.environments.DOCKER.urn,
+                            beam_runner_api_pb2.DockerPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return DockerEnvironment(container_image=payload.container_image)
```

Comment:
there's one problem with this design that still bothers me a little:  the user has to provide the urn in 3 places. 

looking around the code, I see it's pretty common for users to provide the urn at least twice:

```python
  def to_runner_api_parameter(self, context):
    return (common_urns.fixed_windows.urn,
            standard_window_fns_pb2.FixedWindowsPayload(
                size=proto_utils.from_micros(
                    duration_pb2.Duration, self.size.micros),
                offset=proto_utils.from_micros(
                    timestamp_pb2.Timestamp, self.offset.micros)))

  @urns.RunnerApiFn.register_urn(
      common_urns.fixed_windows.urn,
      standard_window_fns_pb2.FixedWindowsPayload)
  def from_runner_api_parameter(fn_parameter, unused_context):
    return FixedWindows(
        size=Duration(micros=fn_parameter.size.ToMicroseconds()),
        offset=Timestamp(micros=fn_parameter.offset.ToMicroseconds()))
```

Environments are different because we have this need to also find the class based on urn.   This makes them more like `Coders`, which have a shortcut for creating the `to_runner_api_parameter` / `from_runner_api_parameter` pair:

```python
class StrUtf8Coder(Coder):
  """A coder used for reading and writing strings as UTF-8."""

  def encode(self, value):
    return value.encode('utf-8')

  def decode(self, value):
    return value.decode('utf-8')

  def is_deterministic(self):
    return True

  def to_type_hint(self):
    return unicode

Coder.register_structured_urn(
    common_urns.coders.STRING_UTF8.urn, StrUtf8Coder)
```

```python
  @staticmethod
  def register_structured_urn(urn, cls):
    # type: (str, Type[Coder]) -> None
    """Register a coder that's completely defined by its urn and its
    component(s), if any, which are passed to construct the instance.
    """
    def to_runner_api_parameter(self, context):
      # type: (Coder, PipelineContext) -> Tuple[str, Any, Sequence[Coder]]
      return (urn, None, self._get_component_coders())

    setattr(cls, 'to_runner_api_parameter', to_runner_api_parameter)

    # pylint: disable=unused-variable
    @Coder.register_urn(urn, None)
    def from_runner_api_parameter(unused_payload, components, unused_context):
      if components:
        return cls(*components)
      else:
        return cls()
```

Let's think a bit more about how we can simplify this while also keeping with the prevailing patterns of the beam code.

Reply:
In this [commit](28c66f78c4d61c6c6cc68dd3c2b97ae15617b2f6) I've attempted to use such pattern for `PythonSDKEnvironment` subclasses, since they fit the pattern most; For other Environments, one thing is that the payload and their constructor signature is different, which makes it harder to fit in